File: lib/tui/hit_map.rb
# frozen_string_literal: true

require_relative "ansi"

module Tui
  # Pure geometry: map a screen (row, col) to a Hit zone using rectangles from
  # a ScreenLayout. Operates on integers and zone symbols only.
  class HitMap
    A = Ansi

    Hit = Data.define(:zone, :payload)

    OUTSIDE = Hit.new(zone: :outside, payload: nil).freeze

    attr_reader :layout, :tab_spans, :row_count, :modal, :popup, :panel,
                :marker_spans, :footer_roles

    def self.build(layout:, tab_spans: [], row_count: 0, modal: nil, popup: nil,
                   panel: false, marker_spans: {}, footer_roles: [])
      new(layout: layout, tab_spans: tab_spans, row_count: row_count,
          modal: modal, popup: popup, panel: panel,
          marker_spans: marker_spans, footer_roles: footer_roles)
    end

    def initialize(layout:, tab_spans:, row_count:, modal:, popup:, panel:,
                   marker_spans:, footer_roles:)
      @layout = layout
      @tab_spans = tab_spans
      @row_count = row_count
      @modal = modal
      @popup = popup
      @panel = panel
      @marker_spans = marker_spans
      @footer_roles = footer_roles
      freeze
    end

    # Never returns nil — unknown coordinates are :outside.
    def at(row, col)
      return OUTSIDE if row.negative? || col.negative?
      return OUTSIDE if row >= @layout.height || col >= @layout.width

      hit_popup(row, col) ||
        hit_modal(row, col) ||
        hit_body(row, col) ||
        hit_header(row, col) ||
        hit_footer(row, col) ||
        hit_border(row, col) ||
        OUTSIDE
    end

    private

    def hit_popup(row, col)
      return unless @popup

      origin_row, origin_col = @layout.body_origin
      pr = @popup[:row]
      pc = @popup[:col]
      lines = @popup[:lines] || []
      return if lines.empty?

      heights = lines.size
      widths = lines.map { |line| A.vislen(line) }
      max_w = widths.max || 0
      screen_row = origin_row + pr
      screen_col = origin_col + pc
      return unless row.between?(screen_row, screen_row + heights - 1)
      return unless col.between?(screen_col, screen_col + max_w - 1)

      local_row = row - screen_row
      line_w = widths[local_row] || 0
      return Hit.new(zone: :border, payload: nil) if col >= screen_col + line_w

      Hit.new(zone: :popup_row, payload: local_row)
    end

    def hit_modal(row, col)
      return unless @modal

      origin_row, origin_col = @layout.body_origin
      placed = @layout.place_modal(@modal)
      return unless placed

      mr = placed[:row]
      mc = placed[:col]
      lines = placed[:lines] || @modal[:lines] || []
      # Mirror Frame.overlay_modal!: box is lines + top/bottom borders, width
      # pinned by modal[:width] or derived the same way Frame does.
      box_h, box_w = modal_box_size(placed, lines)
      screen_row = origin_row + mr
      screen_col = origin_col + mc
      return unless row.between?(screen_row, screen_row + box_h - 1)
      return unless col.between?(screen_col, screen_col + box_w - 1)

      local_row = row - screen_row
      # Interior content rows are 1..(box_h-2); borders are chrome.
      if local_row.positive? && local_row < box_h - 1
        Hit.new(zone: :modal_row, payload: local_row - 1)
      else
        Hit.new(zone: :border, payload: nil)
      end
    end

    def modal_box_size(placed, lines)
      body_h = @layout.body_height
      body_w = @layout.body_width
      if body_h < 3 || body_w < 4
        content = [placed[:title], *lines].compact
        return [[content.size, body_h].min, body_w]
      end

      bw = placed[:width] ||
           [(lines.map { |l| A.vislen(l) }.max || 0),
            A.vislen(placed[:title].to_s) + 6, 30].max + 4
      bw = [[bw, body_w].min, 4].max
      bh = [lines.size, body_h - 2].min + 2
      [bh, bw]
    end

    def hit_body(row, col)
      return unless @layout.body_rows.cover?(row)

      if @panel && @layout.panel?
        divider = @layout.panel_divider_col
        if col == divider
          return Hit.new(zone: :panel_divider, payload: nil)
        end
        if @layout.panel_cols.cover?(col)
          return Hit.new(zone: :panel, payload: row - @layout.body_rows.begin)
        end
      end

      return unless @layout.list_cols.cover?(col)

      vis = row - @layout.body_rows.begin
      abs = @layout.viewport_offset + vis
      return Hit.new(zone: :border, payload: nil) if abs >= @row_count

      # Cursor/prefix columns occupy the first two list cells; marker spans are
      # measured inside row.text (after that prefix).
      text_col = col - @layout.list_cols.begin - 2
      if text_col >= 0 && (span = @marker_spans[abs])
        start_c, end_c = span
        if text_col >= start_c && text_col < end_c
          return Hit.new(zone: :collapse_marker, payload: abs)
        end
      end

      Hit.new(zone: :list_row, payload: abs)
    end

    def hit_header(row, col)
      return unless row == @layout.header_row

      @tab_spans.each do |key, start_col, end_col|
        return Hit.new(zone: :tab, payload: key) if col >= start_col && col < end_col
      end
      Hit.new(zone: :header, payload: nil)
    end

    def hit_footer(row, col)
      return unless @layout.footer_rows.cover?(row)
      return Hit.new(zone: :border, payload: nil) unless col.between?(1, @layout.width - 2)

      index = row - @layout.footer_rows.begin
      role = @footer_roles[index] || :chrome
      Hit.new(zone: :footer_row, payload: { index: index, role: role })
    end

    def hit_border(row, col)
      # Outer ring, header/footer divider rules, and anything else on-frame.
      Hit.new(zone: :border, payload: nil)
    end
  end
end

File: go/internal/tui/term/hitmap/hitmap.go
// Package hitmap is pure geometry: it maps a screen (row, col) to a Hit zone
// using the rectangles a layout publishes. It operates on integers and zone
// names only.
//
// Go port of Ruby's lib/tui/hit_map.rb.
package hitmap

import (
	"tasks-go/internal/tui/term/ansi"
	"tasks-go/internal/tui/term/layout"
)

// Zone names a clickable region of the frame.
type Zone string

const (
	ZoneOutside        Zone = "outside"
	ZoneBorder         Zone = "border"
	ZoneHeader         Zone = "header"
	ZoneTab            Zone = "tab"
	ZoneListRow        Zone = "list_row"
	ZoneCollapseMarker Zone = "collapse_marker"
	ZonePanel          Zone = "panel"
	ZonePanelDivider   Zone = "panel_divider"
	ZoneFooterRow      Zone = "footer_row"
	ZoneModalRow       Zone = "modal_row"
	ZonePopupRow       Zone = "popup_row"
)

// FooterRole classifies a footer line for routing purposes.
type FooterRole string

const (
	RoleChrome   FooterRole = "chrome"
	RoleResponse FooterRole = "response"
	RolePrompt   FooterRole = "prompt"
)

// FooterPayload is the payload a footer hit carries.
type FooterPayload struct {
	Index int
	Role  FooterRole
}

// Hit is one resolved zone. Index carries the row/line payload for list, modal,
// popup and footer zones; Tab carries the tab key; Footer carries the footer
// index and role.
type Hit struct {
	Zone   Zone
	Index  int
	Tab    string
	Footer FooterPayload
}

// Outside is the never-nil answer for coordinates off the frame.
var Outside = Hit{Zone: ZoneOutside}

// TabSpan is one clickable tab: a key and the half-open column range it paints.
type TabSpan struct {
	Key        string
	Start, End int
}

// MarkerSpan is the half-open column range of a row's collapse marker, measured
// inside the row text (after the two-cell cursor prefix).
type MarkerSpan struct{ Start, End int }

// Options builds a HitMap.
type Options struct {
	Layout      *layout.Layout
	TabSpans    []TabSpan
	RowCount    int
	Modal       *layout.Modal
	Popup       *layout.Popup
	Panel       bool
	MarkerSpans map[int]MarkerSpan
	FooterRoles []FooterRole
}

// HitMap answers zone queries for one sampled frame.
type HitMap struct {
	opts Options
}

// Build constructs a hit map.
func Build(opts Options) *HitMap { return &HitMap{opts: opts} }

// At resolves a screen coordinate. It never fails: unknown coordinates are
// ZoneOutside.
func (m *HitMap) At(row, col int) Hit {
	l := m.opts.Layout
	if row < 0 || col < 0 || row >= l.Height || col >= l.Width {
		return Outside
	}
	if hit, ok := m.hitPopup(row, col); ok {
		return hit
	}
	if hit, ok := m.hitModal(row, col); ok {
		return hit
	}
	if hit, ok := m.hitBody(row, col); ok {
		return hit
	}
	if hit, ok := m.hitHeader(row, col); ok {
		return hit
	}
	if hit, ok := m.hitFooter(row, col); ok {
		return hit
	}
	// Outer ring, header/footer divider rules, and anything else on-frame.
	return Hit{Zone: ZoneBorder}
}

func (m *HitMap) hitPopup(row, col int) (Hit, bool) {
	popup := m.opts.Popup
	if popup == nil || len(popup.Lines) == 0 {
		return Hit{}, false
	}
	originRow, originCol := m.opts.Layout.BodyOrigin()
	widths := make([]int, len(popup.Lines))
	maxW := 0
	for i, line := range popup.Lines {
		widths[i] = ansi.VisLen(line)
		if widths[i] > maxW {
			maxW = widths[i]
		}
	}
	screenRow := originRow + popup.Row
	screenCol := originCol + popup.Col
	if row < screenRow || row > screenRow+len(popup.Lines)-1 {
		return Hit{}, false
	}
	if col < screenCol || col > screenCol+maxW-1 {
		return Hit{}, false
	}
	localRow := row - screenRow
	if col >= screenCol+widths[localRow] {
		return Hit{Zone: ZoneBorder}, true
	}
	return Hit{Zone: ZonePopupRow, Index: localRow}, true
}

func (m *HitMap) hitModal(row, col int) (Hit, bool) {
	if m.opts.Modal == nil {
		return Hit{}, false
	}
	l := m.opts.Layout
	originRow, originCol := l.BodyOrigin()
	placed := *m.opts.Modal
	if !placed.Placed {
		placed = l.PlaceModal(placed)
	}
	// Mirror the renderer: the box is lines plus top/bottom borders, its width
	// pinned by Modal.Width or derived the same way the renderer derives it.
	boxW, boxH := m.modalBoxSize(placed)
	screenRow := originRow + placed.Row
	screenCol := originCol + placed.Col
	if row < screenRow || row > screenRow+boxH-1 {
		return Hit{}, false
	}
	if col < screenCol || col > screenCol+boxW-1 {
		return Hit{}, false
	}
	localRow := row - screenRow
	// Interior content rows are 1..(boxH-2); borders are chrome.
	if localRow > 0 && localRow < boxH-1 {
		return Hit{Zone: ZoneModalRow, Index: localRow - 1}, true
	}
	return Hit{Zone: ZoneBorder}, true
}

func (m *HitMap) modalBoxSize(placed layout.Modal) (width, height int) {
	l := m.opts.Layout
	if l.BodyHeight < 3 || l.BodyWidth < 4 {
		content := len(placed.Lines)
		if placed.Title != "" {
			content++
		}
		if content > l.BodyHeight {
			content = l.BodyHeight
		}
		return l.BodyWidth, content
	}
	return l.ModalBoxSize(placed)
}

func (m *HitMap) hitBody(row, col int) (Hit, bool) {
	l := m.opts.Layout
	if !l.BodyRows().Covers(row) {
		return Hit{}, false
	}
	if m.opts.Panel && l.HasPanel() {
		if col == l.PanelDividerCol() {
			return Hit{Zone: ZonePanelDivider}, true
		}
		if l.PanelCols().Covers(col) {
			return Hit{Zone: ZonePanel, Index: row - l.BodyRows().Begin}, true
		}
	}
	if !l.ListCols().Covers(col) {
		return Hit{}, false
	}

	vis := row - l.BodyRows().Begin
	abs := l.ViewportOffset + vis
	if abs >= m.opts.RowCount {
		return Hit{Zone: ZoneBorder}, true
	}

	// Cursor/prefix columns occupy the first two list cells; marker spans are
	// measured inside the row text (after that prefix).
	textCol := col - l.ListCols().Begin - 2
	if textCol >= 0 {
		if span, ok := m.opts.MarkerSpans[abs]; ok {
			if textCol >= span.Start && textCol < span.End {
				return Hit{Zone: ZoneCollapseMarker, Index: abs}, true
			}
		}
	}
	return Hit{Zone: ZoneListRow, Index: abs}, true
}

func (m *HitMap) hitHeader(row, col int) (Hit, bool) {
	l := m.opts.Layout
	if row != l.HeaderRow() {
		return Hit{}, false
	}
	for _, span := range m.opts.TabSpans {
		if col >= span.Start && col < span.End {
			return Hit{Zone: ZoneTab, Tab: span.Key}, true
		}
	}
	return Hit{Zone: ZoneHeader}, true
}

func (m *HitMap) hitFooter(row, col int) (Hit, bool) {
	l := m.opts.Layout
	rows := l.FooterRows()
	if !rows.Covers(row) {
		return Hit{}, false
	}
	if col < 1 || col > l.Width-2 {
		return Hit{Zone: ZoneBorder}, true
	}
	index := row - rows.Begin
	role := RoleChrome
	if index < len(m.opts.FooterRoles) && m.opts.FooterRoles[index] != "" {
		role = m.opts.FooterRoles[index]
	}
	return Hit{Zone: ZoneFooterRow, Footer: FooterPayload{Index: index, Role: role}}, true
}
